### PR TITLE
Mylin/#113 cube histogram [DO NOT merge this PR, may have submodule contamination]

### PR DESCRIPTION
**Pull request by acdo2002**
#113 
A minor discover when I use "socketicd" for developing:
Different servers (socketdev & socketicd) will return different mean(Step8) & stddev(Step9) in "PER_CUBE_HISTOGRAM_HDF5-v15.test.ts". Difference can reach ~0.159, which is much larger than the digit precision we already set.
But the socketdev passes the values which KS designed in the document, and our main server is socketdev as well. So I think it would be ok.
I am just curious about why they the differences? (but I am testing icd15 for long-term performance, the last update dev branch version of socketicd is (4b0f9a8 5068 2020-07-08 18:48:42 UTC))

### CODE REVIEW

**Comment by ajm-asiaa** [comment]
@acdo2002 Maybe the differences between `socketdev` and `socketicd` are that different carta_backends are running behind each, plus they may use different startup parameters? 
I think the `socketdev` carta_backend is always the latest `dev` commit and seems to start up with threads=4 and omp_threads=8 values.
I think you run the carta-backend behind the `socketicd` (port 4013?) yourself?

File: src/test/PER_CUBE_HISTOGRAM-v15.test.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            }, openFileTimeout);
+
+            let ack: AckStream;
+            test(`return RASTER_TILE_DATA(Stream) and check total length `, async () => {
+                await Connection.send(CARTA.AddRequiredTiles, assertItem.addTilesReq);
+                await Connection.send(CARTA.SetCursor, assertItem.setCursor);
+                // await Connection.send(CARTA.SetSpatialRequirements, assertItem.setSpatialReq);
+
+                ack = await Connection.stream(assertItem.addTilesReq.tiles.length + 3) as AckStream;
+                console.log(ack); // RasterTileData * 1 + SpatialProfileData * 1 + RasterTileSync *2 (start & end)
```
Comment (zarda) [inline src/test/PER_CUBE_HISTOGRAM-v15.test.ts:107]: May comment it

File: src/test/PER_CUBE_HISTOGRAM_CANCELLATION-v15.test.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            }, openFileTimeout);
+
+            let ack: AckStream;
+            test(`return RASTER_TILE_DATA(Stream) and check total length `, async () => {
+                await Connection.send(CARTA.AddRequiredTiles, assertItem.addTilesReq);
+                await Connection.send(CARTA.SetCursor, assertItem.setCursor);
+                // await Connection.send(CARTA.SetSpatialRequirements, assertItem.setSpatialReq);
+
+                ack = await Connection.stream(assertItem.addTilesReq.tiles.length + 3) as AckStream;
+                console.log(ack); // RasterTileData * 1 + SpatialProfileData * 1 + RasterTileSync *2 (start & end)
```
Comment (zarda) [inline src/test/PER_CUBE_HISTOGRAM_CANCELLATION-v15.test.ts:116]: May comment it

File: src/test/PER_CUBE_HISTOGRAM_HDF5-v15.test.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            }, openFileTimeout);
+
+            let ack: AckStream;
+            test(`return RASTER_TILE_DATA(Stream) and check total length `, async () => {
+                await Connection.send(CARTA.AddRequiredTiles, assertItem.addTilesReq);
+                await Connection.send(CARTA.SetCursor, assertItem.setCursor);
+                // await Connection.send(CARTA.SetSpatialRequirements, assertItem.setSpatialReq);
+
+                ack = await Connection.stream(assertItem.addTilesReq.tiles.length + 3) as AckStream;
+                console.log(ack); // RasterTileData * 1 + SpatialProfileData * 1 + RasterTileSync *2 (start & end)
```
Comment (zarda) [inline src/test/PER_CUBE_HISTOGRAM_HDF5-v15.test.ts:107]: May comment it

**Comment by acdo2002** [comment]
Hi Anthony @ajm-asiaa ,
Yes I use the carta-backend behind the socketicd (port 4013), with threads=4 and omp_threads=8 (root & base = /scratch/ssd).
`./carta_backend base=/scratch/ssd root=/scratch/ssd port=4013 verbose=true threads=4 omp_threads=8`
But I did not update to the latest dev since last week because I am using socketicd for long-term performance tests.
Perhaps I should keep in mind, and check what happened if I update the socketicd to the latest dev branch.

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +0,0 @@
-[submodule "protobuf"]
```
Comment (zarda) [inline .gitmodules:1]: Submodule should be there.

**Comment by ajm-asiaa** [comment]
@acdo2002 Then in that case, `socketdev` is using /scratch/almalustre (the Lustre disk array), and `socketicd` is using /scratch/ssd (the internal SSD array). 
We think the SSD array should have much faster and consistent performance compared to Lustre.
Is that what your tests show? (`socketicd` faster than `socketdev`)?

**Comment by kswang1029** [comment]
@acdo2002 check hdf5 image in lustre and local ssd. The image might have been updated with new hdf convertor.

**Comment by acdo2002** [comment]
After the tests, I figure out such mean/stdDev differences come from almalustre versus ssd.
If I set root & base to almalustre, `socketicd` will return the same results to `sockedev`

About the speed (execution time)
PER_CUBE_HISTOGRAM-v15.test.ts: 33.09s for almalustre; 33.161s for ssd
PER_CUBE_HISTOGRAM_HDF5-v15.test.ts: 5.085s for almalustre; 1.062s for ssd
PER_CUBE_HISTOGRAM_CANCELLATION-v15.test.ts: 39.183s for almalustre; 38.708s for ssd

hdf5(PER_CUBE_HISTOGRAM_HDF5-v15.test.ts) looks running faster significantly than fits (PER_CUBE_HISTOGRAM-v15.test.ts & PER_CUBE_HISTOGRAM_CANCELLATION-v15.test.ts).

Because this PR may have contamination of submodule (found by Hengtai), please **DO NOT merge this PR.** 
**The clean PR please refer to #115**

**Comment by ajm-asiaa** [comment]
@acdo2002 You might be able to compare the file sizes in bytes between the HDF5 files on Lustre and the SSD. (`ls -sort`). I guess there might be a slight difference in file size if a different converter program was used to create them. 

The latest HDF5 files should be on Lustre because @kswang1029 generated them there. We just later copy them to the SSD.
So you could also check the time stamps. The timestamp on the HDF5 files on the SSD should be more recent than those on Lustre. 
Or in other words, if the timestamp of the HDF5 file on Lustre is more recent than the same file on the SSD, that means it was generated with a new version of the converter, and the HDF5 file on the SSD is old.

**Comment by kswang1029** [comment]
Headers should contain convertor version number

**Comment by ajm-asiaa** [comment]
@acdo2002 Does this still need review or is it finished?

**Comment by acdo2002** [comment]
Hi Anthony @ajm-asiaa ,
This test implementation is finished because the server looks working fine with these codes. 

But I will follow your suggestions to check/investigate the differences in hdf5 latter. I will let everyone knows when I finish the check. ><"
(sorry because there is a contamination issue ><", so I was working on those problems)

**Comment by zarda** [comment]
Merged another polished branch.

**Comment by acdo2002** [comment]
Many thanks Anthony & KS!!!
I have checked the 
`/scratch/images/set_QA`
supermosaic.10.fits; 7G;
supermosaic.10.hdf5.old; 13G; hdf5 converter version: 0.1.4
supermosaic.10.hdf5; 18G; hdf5 converter version: 0.1.10

**Pull request closed without merging**